tfjs-backend-webgpu/src/kernels/FromPixels.ts

```diff
@@ -1,109 +1,187 @@
-  let pixelArray = imageData;
-  if (numChannels != null && numChannels !== 4) {
-    pixelArray = new Uint8Array(pixels.width * pixels.height * numChannels);
-
-    const dataLength = imageData.length;
-    let j = 0;
-    for (let i = 0; i < dataLength; i++) {
-      if (i % 4 < numChannels) {
-        pixelArray[j++] = imageData[i];
-      }
-    }
-  }
-
-  const output = backend.makeOutputArray(outShape, 'int32');
-
-  const info = backend.tensorMap.get(output.dataId);
-  info.values = new Int32Array(pixelArray);
-  backend.maybeReleaseBuffer(output.dataId);
-
-  backend.uploadToGPU(output.dataId);
-  return output as Tensor3D;
-}
+/**
+ * @license
+ * Copyright 2020 Google LLC. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use backend file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * =============================================================================
+ */
+
+import {env, KernelConfig, KernelFunc} from '@tensorflow/tfjs-core';
+import {FromPixels, FromPixelsInputs, FromPixelsAttrs} from '@tensorflow/tfjs-core';
+import {FromPixelsAsync} from '@tensorflow/tfjs-core';
+import {backend_util, Tensor3D} from '@tensorflow/tfjs-core';
+import {WebGPUBackend} from '../backend_webgpu';
+import * as webgpu_program from './webgpu_program';
+import {FromPixelsProgram} from './FromPixels_utils/from_pixels_webgpu';
+import {util} from '@tensorflow/tfjs-core';
+
+export const fromPixelsConfig: KernelConfig = {
+  kernelName: FromPixels,
+  backendName: 'webgpu',
+  kernelFunc: fromPixels as {} as KernelFunc,
+};
+
+// Not many diffs with fromPixel, keep it here
+export const fromPixelsAsyncConfig: KernelConfig = {
+      kernelName: FromPixelsAsync,
+      backendName: 'webgpu',
+      kernelFunc: fromPixelsAsync as {} as KernelFunc,
+    };
+
+let fromPixels2DContext: CanvasRenderingContext2D;
+
+async function fromPixelsAsync(args: {
+      inputs: FromPixelsInputs,
+      backend: WebGPUBackend,
+      attrs: FromPixelsAttrs
+    }) {
+      const { inputs, backend, attrs } = args;
+      const { pixels } = inputs;
+      const { numChannels } = attrs;
+
+      if (pixels == null) {
+        throw new Error(
+          'pixels passed to tf.browser.fromPixels() can not be null');
+      }
+
+      const outShape = [pixels.height, pixels.width, numChannels];
+      const size = util.sizeFromShape(outShape);
+      const uniformData: [number, number] = [size, numChannels];
+      //let imageData = (pixels as ImageData | backend_util.PixelData).data;
```

Comment:
If above line is not needed, remove it.

Reply:
Sry. Force update last CL and it has been covered. This comments has been resolved.